gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb

```diff
@@ -0,0 +1,1078 @@
+        "    )\n",
+        "    retriever = vector_store.as_retriever()\n",
+        "    return str([doc for doc in retriever.invoke(query)])"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 16,
+      "metadata": {
+        "id": "XVjf2fT_bIGa"
+      },
+      "outputs": [],
+      "source": [
+        "@tool\n",
+        "def book_similarity_search(query: str) -> str:\n",
+        "    \"\"\"\n",
+        "    Perform a similarity search for books based on the user's last message.\n",
+        "\n",
+        "    Args:\n",
+        "        state (List[BaseMessage]): The current conversation state, where the last message contains the query.\n",
+        "\n",
+        "    Returns:\n",
+        "        List[BaseMessage]: A list of BaseMessage containing the search results.\n",
+        "    \"\"\"\n",
+        "    engine = PostgresEngine.from_instance(\n",
+        "        PROJECT_ID,\n",
+        "        REGION,\n",
+        "        INSTANCE,\n",
+        "        DATABASE,\n",
+        "        quota_project=PROJECT_ID,\n",
+        "        # Uncomment to use built-in authentication instead of IAM authentication\n",
+        "        user=\"postgres\",\n",
+        "        password=PASSWORD,\n",
+        "    )\n",
+        "\n",
+        "    vector_store = PostgresVectorStore.create_sync(\n",
+        "        engine,\n",
+        "        table_name=BOOK_TABLE_NAME,\n",
+        "        embedding_service=VertexAIEmbeddings(\n",
+        "            model_name=\"textembedding-gecko@latest\", project=PROJECT_ID\n",
+        "        ),\n",
+        "    )\n",
+        "    retriever = vector_store.as_retriever()\n",
+        "    return str([doc for doc in retriever.invoke(query)])"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BF8xqd84351O"
+      },
+      "source": [
+        "### Define router\n",
+        "\n",
+        "Then, you'll define a router to control the flow of the conversation, determining which tool to use based on user input or the state of the interaction. Here we'll use a simple router setup, and you can customize the behavior of your router to handle multiple tools, custom logic, or multi-agent workflows."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 26,
+      "metadata": {
+        "id": "7m9XRSn3351O"
+      },
+      "outputs": [],
+      "source": [
+        "def router(\n",
+        "    state: List[BaseMessage],\n",
+        ") -> Literal[\"book_similarity_search\", \"movie_similarity_search\", \"__end__\"]:\n",
+        "    if not state[0].content or len(state[1].tool_calls) == 0:\n",
+        "        return \"__end__\"\n",
+        "    if \"book\" in state[0].content:\n",
+        "        return \"book_similarity_search\"\n",
+        "    if \"movie\" in state[0].content:\n",
+        "        return \"movie_similarity_search\"\n",
+        "    return \"__end__\""
```

Comment:
Explain in comments or the text cell above how this router works, suggest something along the lines of: "The router will invoke different nodes in the graph based on whether the user prompt contains the word 'book' or 'movie'."

Reply:
Added.